### DISCUSSION TITLE
wire:navigate, alpine's x-for and the browsers back button

### DISCUSSION THREAD

**Discussion by patriktoth67** [livewire/livewire#6849] [7. Bugs (archived)] [answered]
### Livewire version

3.0.5

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

Hi guys, I have a problem with wire:navigate and alpine's x-for loop after clicking the browsers back button.
Here is the simplest code which has the problem, and can be easily replicated.
TEST1 is a full page component, and all links have wire:navigate.

TEST1.php:
```
<?php

namespace App\Livewire\Pages;

use Livewire\Component;

class Test1 extends Component
{
    public function render()
    {
        return view('livewire.pages.test1');
    }
}

```

test1.blade.php:
```
<div class="relative z-20 pt-36" x-data="{
    array: [1, 2],
}">
    <template x-for="arr in array" :key="arr">
        <div x-init="console.log(Math.random());"></div>
    </template>
</div>
```

web.php:
```
Route::get('/test1', Test1::class);
```

The following happens:

1. I open the page (/test1), and it logs this to the console 2 random numbers as expected.

2. I navigate away to another page with wire:navigate link

3. I press the browsers back button

4. Now there are 4 new random numbers logged to the console instead of 2

![image](https://github.com/livewire/livewire/assets/47533464/1dde52e0-7d5f-4338-b7a0-998cbbc976ae)

This issue only appears with alpines x-for directive, with blade's @for it works all good. 

The biggest problem with it is that if I try to use the separate items within the for loop, it throws an error for the 2nd try:

```
<div class="relative z-20 pt-36" x-data="{
    array: [1, 2],
}">
    <template x-for="arr in array" :key="arr">
        <div x-init="console.log(arr);"></div>
    </template>
</div>
```

![image](https://github.com/livewire/livewire/assets/47533464/61584f2e-f6f5-4622-ac9c-93ea3557b5f5)

I'm 99% sure there is a bug with alpines x-for and wire:navigate in livewire, but I'm not qualified to find it.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by aftabkhaliq86** [comment]
also i have  issue x- for when press back  data is null 

```
 <div class="card bottom-0 flex-column-reverse bottom-50" x-show="open"
                                    @click.away="open = false" style="position: absolute;width: 18rem;z-index: 10">
                                    <input type="search" x-model="search" class="outline-none p-1 border border-5" placeholder="search">
                                    <ul class="list-group list-group-flush flex-column-reverse"
                                        @click="focusMessageInput()">
                                        <template x-for="item in searchResults" :key="item.id">
                                            <li class="list-group-item cursor-pointer" x-text="item.name"
                                                x-on:click="cannedMessage(item.message);open = false">
                                            </li>
                                        </template>
                                        <template x-if="searchResults.length === 0">
                                            <li class="list-group-item text-muted text-center">No result found.</li>
                                        </template>
                                    </ul>
                                </div>
```
![image](https://github.com/livewire/livewire/assets/129963183/d34d2e7d-75b5-4e13-97ca-700b74138e3d)

**Reply by patriktoth67** [reply]
Yes, the issue I detected initially was this, but I tested and back traced it to it running everything twice. And on the 2nd run, it'll throw an error if you use the individual items in the for loop.

In your case here:
```
<template x-for="item in searchResults" :key="item.id">
    <li class="list-group-item cursor-pointer" x-text="item.name"
        x-on:click="cannedMessage(item.message);open = false">
    </li>
</template>
```
 You use `item.name` and `item.message` within the x-for template, and that's why you get an error. 
 
For some reason anything you define in the x-for directive as the current iteration element cannot be used within the x-for loop.
 
 It's definitely a livewire bug.

**Reply by danie-ramdhani** [reply]
Yeah, this is the same issue like #6686. but maybe you can do this temporary solution using `x-show`:

```blade
<template x-for="item in searchResults" :key="item.id">
    <li class="list-group-item cursor-pointer" x-show="item" x-cloak x-text="item.name"
        x-on:click="cannedMessage(item.message);open = false">
    </li>
</template>
```

**Comment by N1ebieski** [comment]
Read #5894

You should use an event ```wire:navigating``` and delete all rendered items from x-for before navigate away to another page.

**Reply by patriktoth67** [reply]
Thank you, yes it'll work if I add aclass to the elements, then remove them on the livewire:navigating  event:

```
<div class="relative z-20 pt-36" x-data="{
    array: [1, 2],
    init() {
        document.addEventListener('livewire:navigating', () => {
            let elements = document.querySelectorAll('.remove');
            elements.forEach(el => {
                el.remove();
            });
        });
    },
}">
    <template x-for="arr in array" :key="arr">
        <div class="remove" x-init="console.log(arr)"></div>
    </template>
</div>

```

**Comment by danie-ramdhani** [comment, marked as answer]
fixed https://github.com/alpinejs/alpine/pull/4015 in [v3.4.5](https://github.com/livewire/livewire/releases/tag/v3.4.5)

**Reply by patriktoth67** [reply]
Thank you!
